keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+---
+title: KEP Template
+authors:
+  - "@jennybuckley"
+  - "@apelisse"
+owning-sig: sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-06-21
+last-updated: 2019-06-21
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+---
+
+# Apply Scalability
+
+## Table of Contents
+
+- TODO
+
+## Summary
+
+Serverside Apply (SSA) in its current implementation causes some serious scalability issues when the feature is enabled. Specifically, list operation duration was highly dependent on the number of updates happening. From our testing it seemed to be because of a large number of small object allocations.
+
+## Motivation
+
+### The Problems
+
+From our testing, the problems seem to be because of a large number of small object allocations.
+
+#### On list, update, and apply: 
+
+Deserializing managedFields as a deeply nested field in the apiserver allocates many small objects. Objects are deserialized when fetched from etcd.
+The size of managedFields is similar to the size of the object. It increases etcd size, and cache size of the apiserver and clients.
+
+#### On update and apply: 
+
+Converting managedFields from the api representation into the internal representation causes many small object allocations.
+Some structured-merge-diff operations (like compare) also create many small objects that are not needed. Most “field set” operations are not done in place.
+
+### Experiments
+
+The following experiments convinced us that the problem is mostly the allocations of small objects. Latency in etcd because of read/write with bigger objects, and/or data transfer was not a significant issue.
+
+For all these scenarios, we’ve created a deployment with 5,000 pods. Pods that have the managedFields (when server-side apply is enabled) are obviously larger. Here are some numbers for the comparison, with the objects in protobuf format:
+
+- Server-side apply disabled:
+  - A pod is about 8,137 bytes
```

Comment:
bytes in memory or serialized?  historically pods in memory are 2-10x the size of serialized bytes due mostly to golang maps

Reply:
This is serialized as protobuf

Reply:
(I made pretty big pods)

Reply:
From previous tests with Go 1.6 or 1.7, a single Golang map[string]string was ~360 bytes or so.  So every labels map, annotations, nested map, etc was roughly this expensive.  It might be interesting to know what your API server / controller manager memory usage looked like between scenarios 1 and 2